builtins/target-neon-i32x4.ll

```diff
@@ -258,6 +258,175 @@ define <WIDTH x double> @__sqrt_varying_double(<WIDTH x double>) nounwind readno
+  %r16 = trunc i32 %r to i16
+  ret i16 %r16
+}
+
+define i32 @__reduce_add_int16(<WIDTH x i16>) nounwind readnone alwaysinline {
+  %r = call i32 @llvm.aarch64.neon.saddv.i32.v4i16(<4 x i16> %0)
+  ret i32 %r
+}
+
+define i64 @__reduce_add_int32(<WIDTH x i32>) nounwind readnone alwaysinline {
+  %r = call i32 @llvm.aarch64.neon.saddv.i32.v4i32(<4 x i32> %0)
+  %r64 = sext i32 %r to i64
+  ret i64 %r64
+}
+
+define i32 @__reduce_min_int32(<4 x i32>) nounwind readnone alwaysinline {
+  %r = call i32 @llvm.aarch64.neon.sminv.i32.v4i32(<4 x i32> %0)
+  ret i32 %r
+}
+
+define i32 @__reduce_max_int32(<4 x i32>) nounwind readnone alwaysinline {
+  %r = call i32 @llvm.aarch64.neon.sminv.i32.v4i32(<4 x i32> %0)
+  ret i32 %r
+}
+
+define i32 @__reduce_min_uint32(<4 x i32>) nounwind readnone alwaysinline {
+  %r = call i32 @llvm.aarch64.neon.uminv.i32.v4i32(<4 x i32> %0)
+  ret i32 %r
+}
+
+define i32 @__reduce_max_uint32(<4 x i32>) nounwind readnone alwaysinline {
+  %r = call i32 @llvm.aarch64.neon.uminv.i32.v4i32(<4 x i32> %0)
+  ret i32 %r
+}
+
+define double @__reduce_add_double(<4 x double>) nounwind readnone alwaysinline {
+  %v0 = shufflevector <4 x double> %0, <4 x double> undef,
+                      <2 x i32> <i32 0, i32 1>
+  %v1 = shufflevector <4 x double> %0, <4 x double> undef,
+                      <2 x i32> <i32 2, i32 3>
+  %sum = fadd <2 x double> %v0, %v1
+  %m = call double @llvm.aarch64.neon.faddv.f64.v2f64(<2 x double> %sum)
+  ret double %m
+}
+
+define double @__reduce_min_double(<4 x double>) nounwind readnone alwaysinline {
+  %v0 = shufflevector <4 x double> %0, <4 x double> undef,
+                      <2 x i32> <i32 0, i32 1>
+  %v1 = shufflevector <4 x double> %0, <4 x double> undef,
+                      <2 x i32> <i32 2, i32 3>
+  %x = call <2 x double> @llvm.aarch64.neon.fmin.v2f64(<2 x double> %v0, <2 x double> %v1)
+  %m = call double @llvm.aarch64.neon.fminv.f64.v2f64(<2 x double> %x)
+  ret double %m
+}
+
+define double @__reduce_max_double(<4 x double>) nounwind readnone alwaysinline {
+  %v0 = shufflevector <4 x double> %0, <4 x double> undef,
+                      <2 x i32> <i32 0, i32 1>
+  %v1 = shufflevector <4 x double> %0, <4 x double> undef,
+                      <2 x i32> <i32 2, i32 3>
+  %x = call <2 x double> @llvm.aarch64.neon.fmax.v2f64(<2 x double> %v0, <2 x double> %v1)
+  %m = call double @llvm.aarch64.neon.fmaxv.f64.v2f64(<2 x double> %x)
+  ret double %m
+}
+
+define i64 @__reduce_add_int64(<4 x i64>) nounwind readnone alwaysinline {
+  %v0 = shufflevector <4 x i64> %0, <4 x i64> undef,
+                      <2 x i32> <i32 0, i32 1>
+  %v1 = shufflevector <4 x i64> %0, <4 x i64> undef,
+                      <2 x i32> <i32 2, i32 3>
+  %sum = add <2 x i64> %v0, %v1
+  %m = call i64 @llvm.aarch64.neon.saddv.i64.v2i64(<2 x i64> %sum)
+  ret i64 %m
+}
+
+define i64 @__reduce_min_int64(<4 x i64>) nounwind readnone alwaysinline {
+  reduce4(i64, @__min_varying_int64, @__min_uniform_int64)
```

Comment:
Should these functions that still use `reduce4` not be duplicated?

Reply:
You're right.  I will push a patch to fix these.